### PR TITLE
Gitlab blueprint

### PR DESCRIPTION
**Pull request by simonebruzzechesse**
New Gitlab blueprint.

## Features and open points
- Integrations
    - [x] Identity
        - [x] SAML
    - [ ] Email:
        - [ ] Gmail / Workspace
        - [x] Sendgrid
- [x] ILB
- [ ] MIG
- [ ] Gitaly
- [x] HTTPS SSL
- [x] Gitlab SSH on port 2222
- [x] Check object store, use GCS wherever possible
- [x] Cloud SQL HA
- [ ] Memorystore HA?
- [x] Integration with Cloud Logging for Gitlab running on Docker container

---

I applicable, I acknowledge that I have:
- [x] Read the [contributing guide](https://github.com/GoogleCloudPlatform/cloud-foundation-fabric/blob/master/CONTRIBUTING.md)
- [x] Ran `terraform fmt` on all modified files
- [x] Regenerated the relevant README.md files using [`tools/tfdoc.py`](https://github.com/GoogleCloudPlatform/cloud-foundation-fabric/blob/master/CONTRIBUTING.md#fabric-tools)
- [ ] Made sure all relevant tests pass

### CODE REVIEW

File: blueprints/third-party-solutions/gitlab/ssl.tf [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+locals {
+  cert_subjects = [
+    {
+      country             = "IT"
+      province            = "Lombardy"
+      locality            = "Milan"
```
Comment (juliocc) [inline blueprints/third-party-solutions/gitlab/ssl.tf:22]: well played ;)

File: blueprints/third-party-solutions/gitlab/variables.tf [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+}
+
+variable "cloudsql_config" {
+  description = "Cloud SQL Postgres config."
+  type = object({
+    name             = optional(string, "gitlab-0")
+    database_version = optional(string, "POSTGRES_13")
+    tier             = optional(string, "db-custom-2-8192")
+  })
+  default = {}
```
Comment (juliocc) [inline blueprints/third-party-solutions/gitlab/variables.tf:30]: nullable=false

File: blueprints/third-party-solutions/gitlab/variables.tf [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+}
+
+variable "gcs_config" {
+  description = "GCS for Object Storage config."
+  type = object({
+    enable_versioning = optional(bool, false)
+    location          = optional(string, "EU")
+    storage_class     = optional(string, "STANDARD")
+  })
+  default = {}
```
Comment (juliocc) [inline blueprints/third-party-solutions/gitlab/variables.tf:40]: nullable=false

File: blueprints/third-party-solutions/gitlab/variables.tf [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    }), {})
+    saml = optional(object({
+      forced                 = optional(bool, false)
+      idp_cert_fingerprint   = string
+      sso_target_url         = string
+      name_identifier_format = optional(string, "urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress")
+    }), null)
+    ha_required = optional(bool, false)
+  })
+  default = {}
```
Comment (juliocc) [inline blueprints/third-party-solutions/gitlab/variables.tf:63]: nullable=false

File: blueprints/third-party-solutions/gitlab/variables.tf [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+      size = optional(number, 20)
+      type = optional(string, "pd-standard")
+    }), {})
+    data_disk = optional(object({
+      size         = optional(number, 100)
+      type         = optional(string, "pd-ssd")
+      replica_zone = optional(string, "europe-west8-c")
+    }), {})
+  })
+}
```
Comment (juliocc) [inline blueprints/third-party-solutions/gitlab/variables.tf:84]: Remove defalut regions

File: blueprints/third-party-solutions/gitlab/variables.tf [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+  description = "Redis Config."
+  type = object({
+    memory_size_gb      = optional(number, 1)
+    name                = optional(string, "gitlab-0")
+    persistence_mode    = optional(string, "RDB")
+    rdb_snapshot_period = optional(string, "TWELVE_HOURS")
+    tier                = optional(string, "BASIC")
+    version             = optional(string, "REDIS_6_X")
+  })
+  default = {}
```
Comment (juliocc) [inline blueprints/third-party-solutions/gitlab/variables.tf:129]: nullable=false

File: blueprints/third-party-solutions/gitlab/variables.tf [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+}
+
+variable "region" {
+  description = "GCP Region."
+  type        = string
+  default     = "europe-west1"
```
Comment (juliocc) [inline blueprints/third-party-solutions/gitlab/variables.tf:135]: remove default region

**Review by juliocc — APPROVED** [review]
very nice work.

Left a few comments below.

File: blueprints/third-party-solutions/README.md [hunk 8]
```diff
@@ -26,4 +26,10 @@ The blueprints in this folder show how to automate installation of specific thir
 
 <a href="./f5-bigip/" title="F5 BigIP"><img src="./phpipam/diagram.png" align="left" width="320px"></a> <p style="margin-left: 340px">These examples show how to deploy F5 BigIP-VE load balancers in GCP.</p>
 
+<br clear="left">
+
+### Gitlab
+
+<a href="./gitlab/" title="Gitlab"><img src="./gitlab/diagram.png" align="left" width="320px"></a> <p style="margin-left: 340px">This blueprint show how to deploy a Gitlab instance in GCP. The architecture is based on the reference described in the official documentation available at this [link](https://docs.gitlab.com/ee/administration/reference_architectures/1k_users.html) with managed services such as Cloud SQL, Memorystore and Cloud Storage.</p>
```
[the remark this answers is not in the document]
Comment (simonebruzzechesse) [reply blueprints/third-party-solutions/README.md]: Updated

File: blueprints/third-party-solutions/gitlab/gitlab.tf [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+  ]
+  network_interfaces = [
+    {
+      network    = var.network_config.network_self_link
+      subnetwork = var.network_config.subnet_self_link
+    }
+  ]
+  tags = var.gitlab_instance_config.network_tags
+  metadata = {
+    user-data = templatefile("${path.module}/assets/cloud-config.yaml", {
```
[the remark this answers is not in the document]
Comment (simonebruzzechesse) [reply blueprints/third-party-solutions/gitlab/gitlab.tf]: Done!

File: blueprints/third-party-solutions/gitlab/README.md [hunk 10]
```diff
@@ -0,0 +1,350 @@
...
+
+| name | description | sensitive | consumers |
+|---|---|:---:|---|
+| [gitlab_ilb_ip](outputs.tf#L26) | Gitlab Internal Load Balancer IP Address. |  |  |
+| [instance](outputs.tf#L31) | Gitlab compute engine instance. |  |  |
+| [postgresql_users](outputs.tf#L36) | Gitlab postgres user password. | ✓ |  |
+| [project](outputs.tf#L42) | GCP project. |  |  |
+| [ssh_to_gitlab](outputs.tf#L47) | gcloud command to ssh gitlab instance. |  |  |
+| [ssl_certs](outputs.tf#L52) | Gitlab SSL Certificates. |  |  |
+<!-- END TFDOC -->
```
[the remark this answers is not in the document]
Comment (simonebruzzechesse) [reply blueprints/third-party-solutions/gitlab/README.md:350]: Done!

**Pull request merged**